src/Tests/Microsoft.Diagnostics.Monitoring.UnitTests/AuthenticationTests.cs

```diff
@@ -0,0 +1,232 @@
+            var processes = await client.GetProcessesAsync(DefaultApiTimeout);
+            Assert.NotNull(processes);
+
+            // Check that /metrics does not challenge for authentication
+            var metrics = await client.GetMetricsAsync(DefaultApiTimeout);
+            Assert.NotNull(metrics);
+        }
+
+        /// <summary>
+        /// Tests that API key authentication can be configured correctly and
+        /// that the key can be rotated wihtout shutting down dotnet-monitor.
+        /// </summary>
+        [Fact]
+        public async Task ApiKeyAuthenticationSchemeTest()
+        {
+            const string AlgorithmName = "SHA256";
+
+            await using DotNetMonitorRunner toolRunner = new DotNetMonitorRunner(_outputHelper);
+
+            _outputHelper.WriteLine("Generating API key.");
+
+            // Generate initial API key
+            byte[] apiKey = GenerateApiKey();
+
+            // Set API key via key-per-file
+            RootOptions options = new();
+            options.UseApiKey(AlgorithmName, apiKey);
+            toolRunner.WriteKeyPerValueConfiguration(options);
+
+            // Start dotnet-monitor
+            await toolRunner.StartAsync(DefaultStartTimeout);
+
+            // Create HttpClient with default request headers
+            using HttpClient httpClient = new HttpClient();
+            httpClient.DefaultRequestHeaders.Authorization =
+                new AuthenticationHeaderValue(ApiKeyScheme, Convert.ToBase64String(apiKey));
+
+            using ApiClient client = new ApiClient(_outputHelper, await toolRunner.GetDefaultAddressAsync(DefaultAddressTimeout), httpClient);
+
+            // Check that /processes does not challenge for authentication
+            var processes = await client.GetProcessesAsync(DefaultApiTimeout);
+            Assert.NotNull(processes);
+
+            _outputHelper.WriteLine("Rotating API key.");
+
+            // Rotate the API key
+            byte[] apiKey2 = GenerateApiKey();
+
+            options.UseApiKey(AlgorithmName, apiKey2);
+            toolRunner.WriteKeyPerValueConfiguration(options);
+
+            // Wait for the key rotation to be consumed by dotnet-monitor; detect this
+            // by checking for when API returns a 401. Ideally, key rotation would write
+            // log event and runner monitor for event and notify.
+            int attempts = 0;
+            while (true)
+            {
+                attempts++;
+                _outputHelper.WriteLine("Waiting for key rotation (attempt #{0}).", attempts);
+
+                try
+                {
+                    await client.GetProcessesAsync(DefaultApiTimeout);
+                }
+                catch (ApiStatusCodeException ex) when (ex.StatusCode == HttpStatusCode.Unauthorized)
+                {
+                    break;
+                }
+
+                await Task.Delay(TimeSpan.FromSeconds(3));
```

Comment:
nit: Consider delay followed by attempt. I'm guessing it never succeeds on the first try.

Reply:
It typically does not succeed on the first try. However, I plan to replace this with some log listening code once #74 is completed. I will switch the order as you requested.